elastic/kibana
# [Index management] Server-side NP ready

This is the first PR of 3 to migrate Index management to the new platform to make it easier to test and review.

This PR migrates the **server-side** code  and update all the API routes to

- Use the new platform router (with schema validation)
- Use the "licensing" new platform plugin to guard our API route with license check

### How to test

- Navigate through the app and make sure that all requests are fulfilled correctly.
- Open the flyout details and make sure all tabs display the correct data
- Test all index drop-down actions from the list of indices (not on `.kibana-1` index! 😄 )
- Create / edit / delete an index template
- Make sure the extension points from ILM, Rollup and CCR are correctly added. You can see it on the request reponse, each index should have the metadata added

<img width="502" alt="Screen Shot 2020-02-05 at 14 42 41" src="https://user-images.githubusercontent.com/2854616/73828814-f3d57b00-4827-11ea-8abd-b861ecb53c68.png">

And for follower indices a badge should appear in the table

<img width="883" alt="Screen Shot 2020-02-05 at 14 42 05" src="https://user-images.githubusercontent.com/2854616/73828854-018b0080-4828-11ea-99d8-01a2c3af5165.png">


### Code structure proposal

I am making the proposal to put the logic inside the plugin `setup()` phase inside **dedicated services**. The end result should be something like this

<img width="688" alt="Screen Shot 2020-02-05 at 14 47 23" src="https://user-images.githubusercontent.com/2854616/73827960-7826fe80-4826-11ea-84cb-56f55c68d791.png">

As you can see both the route registering and the license are encapsulated inside their own classes. Which I think is cleaner than what we currently have with `watcher`.

<img width="714" alt="Screen Shot 2020-02-05 at 14 48 50" src="https://user-images.githubusercontent.com/2854616/73828089-ae647e00-4826-11ea-8c44-17914796349e.png">

Once the whole migration of Index management is done (moved to the "plugins" folder), I think it would be good to move the `License.ts` service out to our "es-ui-shared" folder and make it available to all our apps (through the np dependencies mechanism).

cc @cjcenizal @jloleysens @alisonelizabeth
